File: Java/LeetCode/Count the Number of Special Characters II.java
// https://leetcode.com/problems/count-the-number-of-special-characters-ii/

class Solution {

    public int numberOfSpecialChars(String word) {
        Map<Character, Integer> lowercaseLastIndices = new HashMap<>();
        Map<Character, Integer> uppercaseFirstIndices = new HashMap<>();

        for (int i = 0; i < word.length(); ++i) {
            char c = word.charAt(i);
            if (Character.isLowerCase(c)) {
                lowercaseLastIndices.put(c, i);
            } else {
                uppercaseFirstIndices.putIfAbsent(c, i);
            }
        }

        int specialCount = 0;
        for (Map.Entry<Character, Integer> entry : lowercaseLastIndices.entrySet()) {
            Character uppercaseKey = Character.toUpperCase(entry.getKey());
            if (uppercaseFirstIndices.getOrDefault(uppercaseKey, -1) > entry.getValue()) {
                specialCount++;
            }
        }

        return specialCount;
    }

}

File: Kotlin/LeetCode/Count the Number of Special Characters II.kt
// https://leetcode.com/problems/count-the-number-of-special-characters-ii/

class Solution {

    fun numberOfSpecialChars(word: String): Int {
        val lowercaseLastIndices = mutableMapOf<Char, Int>();
        val uppercaseFirstIndices = mutableMapOf<Char, Int>();

        word.forEachIndexed { i, c ->
            if (c.isLowerCase()) {
                lowercaseLastIndices.put(c, i);
            } else {
                uppercaseFirstIndices.putIfAbsent(c, i);
            }
        }

        var specialCount = 0;
        lowercaseLastIndices.forEach {
            if (uppercaseFirstIndices.getOrDefault(it.key.toUpperCase(), -1) > it.value) {
                specialCount++;
            }
        }

        return specialCount
    }

}
